pkg/service/controller.go

```diff
@@ -36,72 +37,136 @@ var ControllerCaps = []csi.ControllerServiceCapability_RPC_Type{
+			return nil, fmt.Errorf("failed building disk resource, error is %w", err)
+		}
+		correlationID := fmt.Sprintf("create_disk_%s", utilrand.String(5))
+		createDisk, err := conn.SystemService().DisksService().
+			Add().
+			Disk(disk).
+			Query("correlation_id", correlationID).
+			Send()
+		if err != nil {
+			msg := fmt.Errorf("failed creating disk %s error is %w", diskName, err)
+			klog.Errorf(msg.Error())
+			return nil, msg
+		}
+		disk = createDisk.MustDisk()
+		finished, err := checkJobFinished(ctx, conn, correlationID)
+		if err != nil {
+			msg := fmt.Errorf("error while waiting for disk %s create job to finish, error: %w",
+				disk.MustId(), err)
+			klog.Errorf(msg.Error())
+			return nil, msg
+		}
+		if !finished {
+			msg := fmt.Errorf("create disk %s job didn't finish, error: %w", disk.MustId(), err)
+			klog.Errorf(msg.Error())
+			return nil, msg
+		}
 	}
-
-	// TODO rgolan the default in case of error would be non thin - change it?
-	thinProvisioning, _ := strconv.ParseBool(req.Parameters[ParameterThinProvisioning])
-
-	provisionedSize := req.CapacityRange.GetRequiredBytes()
-	if provisionedSize < minimumDiskSize {
-		provisionedSize = minimumDiskSize
-	}
-
-	// creating the disk
-	disk, err := ovirtsdk.NewDiskBuilder().
-		Name(req.Name).
-		StorageDomainsBuilderOfAny(*ovirtsdk.NewStorageDomainBuilder().Name(req.Parameters[ParameterStorageDomainName])).
-		ProvisionedSize(provisionedSize).
-		ReadOnly(false).
-		Format(ovirtsdk.DISKFORMAT_COW).
-		Sparse(thinProvisioning).
-		Build()
-
-	if err != nil {
-		// failed to construct the disk
-		return nil, err
-	}
-
-	createDisk, err := conn.SystemService().DisksService().
-		Add().
-		Disk(disk).
-		Send()
-	if err != nil {
-		// failed to create the disk
-		klog.Errorf("Failed creating disk %s", req.Name)
-		return nil, err
+	if err = waitForDiskStatusOk(ctx, conn, disk.MustId()); err != nil {
+		msg := fmt.Errorf("failed waiting for disk %s to be created, error: %w",
+			disk.MustId(), err)
+		klog.Errorf(msg.Error())
+		return nil, msg
 	}
+	klog.Infof("Finished creating disk %s", diskName)
 	return &csi.CreateVolumeResponse{
 		Volume: &csi.Volume{
-			CapacityBytes: createDisk.MustDisk().MustProvisionedSize(),
-			VolumeId:      createDisk.MustDisk().MustId(),
+			CapacityBytes: disk.MustProvisionedSize(),
+			VolumeId:      disk.MustId(),
 		},
 	}, nil
 }
 
+func handleCreateVolumeImageFormat(conn *ovirtsdk.Connection, storageDomainName string, thinProvisioning bool) (ovirtsdk.DiskFormat, error) {
+	sd, err := getStorageDomainByName(conn, storageDomainName)
+	if err != nil {
+		return "", fmt.Errorf(
+			"failed searching for storage domain with name %s, error: %w", storageDomainName, err)
+	}
+	if sd == nil {
+		return "", fmt.Errorf(
+			"storage domain with name %s wasn't found", storageDomainName)
+	}
+	storage, ok := sd.Storage()
+	if !ok {
+		return "", fmt.Errorf(
+			"storage domain with name %s didn't have host storage, veify it is connected to a host",
```

Comment:
what's host storage?
I think this branch means there's a bug in the engine or someone edited the database directly

Reply:
Yes probably a bug in the engine but we aim to check all cases, if the port to the go-ovirt-client we have a special error to report something which means a bug in the engine (probably), when we will migrate the CSI driver to use it this will make it easier.
Anyway, we instead to check as many return values as possible.

Reply:
regarding the Host storage, I have no idea this is the name of the type in the go-client:
![image](https://user-images.githubusercontent.com/38602062/127298906-c2bccf09-2040-4039-958c-8a1239e514c8.png)

Feel free to suggest a better name :)

Reply:
small typo: veify = verify